tests/rptest/tests/control_character_flag_test.py

```diff
@@ -0,0 +1,195 @@
+
+    def _validate_pre_upgrade(self, version):
+        self._installer.install(self.redpanda.nodes, version)
+        super(ControlCharacterPermittedBase, self).setUp()
+
+        unique_versions = wait_for_num_versions(self.redpanda, 1)
+        assert ver_string(version) in unique_versions
+
+        admin = Admin(self.redpanda)
+        config = admin.get_cluster_config()
+        assert config.keys().isdisjoint(
+            {self.feature_legacy_permit_control_char})
+
+    def _perform_update(self, initial_version, version):
+        self._installer.install(self.redpanda.nodes, version)
+        self.redpanda.restart_nodes(self.redpanda.nodes)
+
+        unique_versions = wait_for_num_versions(self.redpanda, 1)
+        assert ver_string(initial_version) not in unique_versions
+
+        admin = Admin(self.redpanda)
+        config = admin.get_cluster_config()
+        assert config.keys() > {self.feature_legacy_permit_control_char}
+
+        assert config[self.feature_legacy_permit_control_char] == True
+
+
+class ControlCharacterPermittedAfterUpgrade(ControlCharacterPermittedBase):
+    """
+    When upgrading to Redpanda v23.2.1, a new flag is added that will
+    permit the use of strings that have control characters in them.  When this
+    flag is set to true, these characters are permitted.  When false, the characters
+    are rejected.  In new clusters, those that start at v23.2.1, the flag does
+    nothing.
+    """
+    def __init__(self, test_context):
+        super(ControlCharacterPermittedAfterUpgrade,
+              self).__init__(test_context=test_context, num_brokers=3)
+
+    @cluster(num_nodes=3, log_allow_list=RESTART_LOG_ALLOW_LIST)
+    @parametrize(initial_version=(22, 2, 9))
+    @parametrize(initial_version=(22, 3, 11))
+    @parametrize(initial_version=(23, 1, 1))
+    def test_upgrade_from_pre_v23_2(self, initial_version):
+        """
+        Validate that when upgrading from pre v23.2 that the flag is
+        honored
+        """
+        self._validate_pre_upgrade(initial_version)
+
+        # Creates a user with invalid control characters
+        admin = Admin(self.redpanda)
+        admin.create_user("my\nuser", "password", "SCRAM-SHA-256")
+        self._perform_update(initial_version, RedpandaInstaller.HEAD)
+        # Should still be able to create a user
+        admin.create_user("my\notheruser", "password", "SCRAM-SHA-256")
+        admin.patch_cluster_config(
+            {self.feature_legacy_permit_control_char: False})
+        try:
+            admin.create_user("my\nthirduser", "password", "SCRAM-SHA-256")
+            assert False
+        except requests.exceptions.HTTPError:
+            pass
+
+    @cluster(num_nodes=3, log_allow_list=RESTART_LOG_ALLOW_LIST)
+    def test_new_cluster(self):
+        """
+        Validates that new clusters ignore the flag
+        """
+        super(ControlCharacterPermittedAfterUpgrade, self)._start_redpanda()
```

Comment:
The `ControlCharacterPermittedBase` class, where all the test classes inherit from, overrides `setUp()` to not do anything.  This is so I can prevent Redpanda from starting until I have selected which version I want to start with.  Once I have selected the installation target (because I want to start with an older cluster), then I start Redpanda.